### DISCUSSION TITLE
SVG Scratch Blocks For Costume Editor?

### DISCUSSION THREAD

**Discussion by StickmanRed** [ScratchAddons/ScratchAddons#7321] [Ideas] [answered]
When importing SVG Scratch blocks obtained through the "Save blocks as image" addon into the Scratch costume editor, the text becomes misaligned... would a feature that adjusts the text somewhere in between exporting and importing be useful enough to be added to the extension? Or would that fall into a broader addon that fixes all SVG text issues, as proposed in [this discussion](https://github.com/ScratchAddons/ScratchAddons/discussions/7084)? (Right now, I have to run the adjustment code and import it manually...)

**Comment by WorldLanguages** [comment]
> (Right now, I have to run the adjustment code and import it manually...)

What's the "adjustment code"?

**Reply by StickmanRed** [reply]
> > (Right now, I have to run the adjustment code and import it manually...)
> 
> What's the "adjustment code"?

Sorry, I meant to say that I have to use another code to manually adjust the exported SVGs for the Scratch editor to display them properly...

**Reply by WorldLanguages** [reply]
And could you tell us what code you use?

**Reply by StickmanRed** [reply]
```javascript
const texts = document.getElementsByTagName("text");
var attr;
for (i of texts) {
    i.setAttribute("text-anchor", "start");
    try {attr = i.getAttribute("transform").substr(10).replaceAll(")","").trimEnd().split(", ");}
    catch {attr = ["0","0"];}
    attr[0] = Number.parseFloat(attr[0]); attr[1] = Number.parseFloat(attr[1]);
    if (i.parentElement.getAttribute("class") === "blocklyEditableText") {
        i.setAttribute("transform", `translate(${attr[0] + 3}, ${attr[1] + 2}) `);
    }
    else {
        i.setAttribute("transform", `translate(${attr[0]}, ${attr[1] - 14}) `);
    }
}
```

it only accounts for the text displacement in SVGs specifically obtained through the "Save blocks as image" addon (or the Scratch code editor)...

**Reply by WorldLanguages** [reply, marked as answer]
We could add that logic to the `SVG uploads fix` addon, if you give us permission.

**Reply by StickmanRed** [reply]
actually, I think the below would fix most text errors... the previous code only applied to those specific SVGs, and the text misalignment problem had nothing to do with "text-anchor"

```javascript
const svg = document; // Replace document with the file
const svgWindow = svg.defaultView;

var translate, x, y, size;

for (i of svg.getElementsByTagName("text")) {
  try {
    translate = i
      .getAttribute("transform")
      .substr(10)
      .replaceAll(")", "")
      .trimEnd()
      .split(", ");
  } catch {
    translate = ["0", "0"];
  }
  translate[0] = Number.parseFloat(translate[0]);
  translate[1] = Number.parseFloat(translate[1]);

  try {
    x = i.getAttribute("x");
  } catch {
    x = "0";
  }
  x = Number.parseFloat(x);
  
  try {
    y = i.getAttribute("y");
  } catch {
    y = "0";
  }
  y = Number.parseFloat(y);

  size = svgWindow.getComputedStyle(i).getPropertyValue("font-size");
  size = Number.parseFloat(size.slice(0, -2));

  i.setAttribute(
    "transform",
    `translate(${translate[0] + x}, ${translate[1] + y - size}) `,
  );
  i.setAttribute("x", "0");
  i.setAttribute("y", "0");
}
```
